### PR TITLE
SONARJAVA-1091-2

### PR DESCRIPTION
**Pull request by benzonico**
Fix according to @Godin review.

I will let more time before merging it this time, promise.

### CODE REVIEW

File: java-jacoco/src/main/java/org/sonar/plugins/jacoco/JaCoCoReportMerger.java [hunk 1]
```diff
@@ -64,23 +63,22 @@ public static void mergeReports(File reportOverall, File... reports) {
...
-          isCurrentVersionFormat = AbstractAnalyzer.readJacocoReport(report, dataStore, infoStore);
+          JacocoReportReader jacocoReportReader = new JacocoReportReader(report).readJacocoReport(dataStore, infoStore);
+          isCurrentVersionFormat = jacocoReportReader.useCurrentBinaryFormat();
         } catch (IOException e) {
           throw new SonarException(String.format("Unable to read %s", report.getAbsolutePath()), e);
         }
       }
     }
-    return isCurrentVersionFormat;
+    return BooleanUtils.isNotFalse(isCurrentVersionFormat);
```
Comment (Godin) [inline java-jacoco/src/main/java/org/sonar/plugins/jacoco/JaCoCoReportMerger.java]: @benzonico I don't see that https://github.com/SonarSource/sonar-java/pull/242#discussion_r31311659 was addressed here?

Comment (Godin) [inline java-jacoco/src/main/java/org/sonar/plugins/jacoco/JaCoCoReportMerger.java]: I probably miss something as test case seems to be in place, but then code isn't obvious.

Comment (benzonico) [inline java-jacoco/src/main/java/org/sonar/plugins/jacoco/JaCoCoReportMerger.java]: In fact, I thought JaCoCo was doing some enforcement about this but this is not at all the case, my unit test is flawed. I'll fix this.

Comment (benzonico) [inline java-jacoco/src/main/java/org/sonar/plugins/jacoco/JaCoCoReportMerger.java]: @Godin fix in https://github.com/SonarSource/sonar-java/commit/b3457b80f341ee0dc2eb3aebc2e9ffcc44a913fa

File: java-jacoco/src/main/java/org/sonar/plugins/jacoco/AbstractAnalyzer.java [hunk 2]
```diff
@@ -143,22 +134,20 @@ private void populateClassFilesCache(File dir, String path) {
...
-        useCurrentFormat = readJacocoReport(jacocoExecutionData, executionDataVisitor);
-      } catch (IOException e) {
-        throw new SonarException(e);
-      }
+    }
+    ExecutionDataVisitor executionDataVisitor = new ExecutionDataVisitor();
+    try {
+      jacocoReportReader = new JacocoReportReader(jacocoExecutionData).readJacocoReport(executionDataVisitor, executionDataVisitor);
+    } catch (IOException e) {
+      throw new SonarException("An error occured while reading JaCoCo report", e);
```
Comment (Godin) [inline java-jacoco/src/main/java/org/sonar/plugins/jacoco/AbstractAnalyzer.java]: @benzonico once again: error message about failure to read file is inconsistent - see JaCoCoReportMerger. Code can be simplified, if you'll pull catch of IOException down into JacocoReportReader - less places, where exception should be handled; single place, where message will be constructed.

File: java-jacoco/src/main/java/org/sonar/plugins/jacoco/JaCoCoReportMerger.java [hunk 3]
```diff
@@ -64,23 +63,23 @@ public static void mergeReports(File reportOverall, File... reports) {
...
       if (report.isFile()) {
-        try {
-          isCurrentVersionFormat = AbstractAnalyzer.readJacocoReport(report, dataStore, infoStore);
-        } catch (IOException e) {
-          throw new SonarException(String.format("Unable to read %s", report.getAbsolutePath()), e);
+        JacocoReportReader jacocoReportReader = new JacocoReportReader(report).readJacocoReport(dataStore, infoStore);
+        boolean reportFormatIsCurrent = jacocoReportReader.useCurrentBinaryFormat();
+        if (isCurrentVersionFormat == null) {
+          isCurrentVersionFormat = reportFormatIsCurrent;
+        } else if (isCurrentVersionFormat != reportFormatIsCurrent) {
```
Comment (SonarTech) [inline java-jacoco/src/main/java/org/sonar/plugins/jacoco/JaCoCoReportMerger.java]: MAJOR Change this comparison to use the equals method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 new issue:
* 1 major

**Comment by Godin** [comment]
@benzonico in order to increase sharing of knowledge within team, I'd  like to propose that final review/merge will be done by @m-g-sonar ;)

**Comment by benzonico** [comment]
So it is assigned to him ! :)

**Comment by m-g-sonar** [comment]
LGTM!

**Pull request merged**